// @designliquido/potigol/bibliotecas/biblioteca-global.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.abs = abs;
exports.aleatorio = aleatorio;
exports.arccos = arccos;
exports.arcsen = arcsen;
exports.arctg = arctg;
exports.cos = cos;
exports.log = log;
exports.log10 = log10;
exports.pi = pi;
exports.raiz = raiz;
exports.sen = sen;
exports.tg = tg;
async function abs(interpretador, valor) {
    return Promise.resolve(Math.abs(valor));
}
async function aleatorio() {
    return Promise.resolve(Math.random());
}
async function arccos(interpretador, valor) {
    return Promise.resolve(Math.acos(valor));
}
async function arcsen(interpretador, valor) {
    return Promise.resolve(Math.asin(valor));
}
async function arctg(interpretador, valor) {
    return Promise.resolve(Math.atan(valor));
}
async function cos(interpretador, valor) {
    return Promise.resolve(Math.cos(valor));
}
async function log(interpretador, valor) {
    return Promise.resolve(Math.log(valor));
}
async function log10(interpretador, valor) {
    return Promise.resolve(Math.log10(valor));
}
async function pi() {
    return Promise.resolve(Math.PI);
}
async function raiz(interpretador, valor) {
    return Promise.resolve(Math.sqrt(valor));
}
async function sen(interpretador, valor) {
    return Promise.resolve(Math.sin(valor));
}
async function tg(interpretador, valor) {
    return Promise.resolve(Math.tan(valor));
}
