// @designliquido/potigol/bibliotecas/primitivas-numero.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = {
    arredonde: (interpretador, numero) => Promise.resolve(Math.ceil(numero)),
    caractere: (interpretador, numero) => Promise.resolve(String.fromCharCode(numero)),
    inteiro: (interpretador, numero) => Promise.resolve(Math.floor(numero)),
    formato: (interpretador, numero, mascara) => {
        if (mascara.includes('f')) {
            const casasDecimais = mascara.match(/%\.(\d+)f/)[1];
            return Promise.resolve(numero.toFixed(parseInt(casasDecimais)));
        }
        const casasInteiras = mascara.match(/%(\d+)d/)[1];
        return Promise.resolve(String(numero).padStart(parseInt(casasInteiras), ' '));
    },
    qual_tipo: (interpretador, numero) => Promise.resolve(Math.floor(numero) === numero ? 'Inteiro' : 'Real'),
    piso: (interpretador, numero) => Promise.resolve(Math.floor(numero)),
    real: (interpretador, numero) => Promise.resolve(numero),
    teto: (interpretador, numero) => Promise.resolve(Math.ceil(numero)),
    texto: (interpretador, numero) => Promise.resolve(String(numero)),
};
