// @designliquido/portugol-studio/bibliotecas/calendario.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.dia_mes_atual = dia_mes_atual;
exports.dia_semana_atual = dia_semana_atual;
exports.mes_atual = mes_atual;
exports.ano_atual = ano_atual;
exports.hora_atual = hora_atual;
exports.minuto_atual = minuto_atual;
exports.segundo_atual = segundo_atual;
exports.milisegundo_atual = milisegundo_atual;
exports.dia_semana_completo = dia_semana_completo;
exports.dia_semana_curto = dia_semana_curto;
exports.dia_semana_abreviado = dia_semana_abreviado;
async function dia_mes_atual() {
    const data = new Date();
    return data.getDate();
}
async function dia_semana_atual() {
    const data = new Date();
    let diaDaSemana = data.getDay() + 1;
    if (diaDaSemana === 7)
        diaDaSemana = 0;
    return diaDaSemana;
}
async function mes_atual() {
    const data = new Date();
    return data.getMonth() + 1;
}
async function ano_atual() {
    const data = new Date();
    return data.getFullYear();
}
async function hora_atual(interpretador, formato_12h) {
    const data = new Date();
    if (!formato_12h) {
        return data.getHours();
    }
    let hora = data.getHours() % 12;
    if (hora === 0) {
        hora = 12;
    }
    return hora;
}
async function minuto_atual() {
    const data = new Date();
    return data.getMinutes();
}
async function segundo_atual() {
    const data = new Date();
    return data.getSeconds();
}
async function milisegundo_atual() {
    const data = new Date();
    return data.getMilliseconds();
}
async function dia_semana_completo(interpretador, numero_dia, caixa_alta, caixa_baixa) {
    const dias = [
        'Domingo',
        'Segunda-Feira',
        'Terça-Feira',
        'Quarta-Feira',
        'Quinta-Feira',
        'Sexta-Feira',
        'Sábado',
    ];
    if (numero_dia > 0 && numero_dia < 8) {
        let dia = dias[numero_dia - 1];
        if (caixa_alta) {
            dia = dia.toUpperCase();
        }
        else if (caixa_baixa) {
            dia = dia.toLowerCase();
        }
        return dia;
    }
    throw new Error(`'${numero_dia}' não corresponde a um dia da semana válido.`);
}
async function dia_semana_curto(interpretador, numero_dia, caixa_alta, caixa_baixa) {
    const dias = ['Domingo', 'Segunda', 'Terça', 'Quarta', 'Quinta', 'Sexta', 'Sabado'];
    if (numero_dia > 0 && numero_dia < 8) {
        let dia = dias[numero_dia - 1];
        if (caixa_alta) {
            dia = dia.toUpperCase();
        }
        else if (caixa_baixa) {
            dia = dia.toLowerCase();
        }
        return dia;
    }
    throw new Error(`'${numero_dia}' não corresponde a um dia da semana válido.`);
}
async function dia_semana_abreviado(interpretador, numero_dia, caixa_alta, caixa_baixa) {
    const dias = ['Dom', 'Seg', 'Ter', 'Qua', 'Qui', 'Sex', 'Sab'];
    if (numero_dia > 0 && numero_dia < 8) {
        let dia = dias[numero_dia - 1];
        if (caixa_alta) {
            dia = dia.toUpperCase();
        }
        else if (caixa_baixa) {
            dia = dia.toLowerCase();
        }
        return dia;
    }
    throw new Error(`'${numero_dia}' não corresponde a um dia da semana válido.`);
}
