// @designliquido/delegua-node/nucleo-traducao.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.NucleoTraducao = void 0;
const sistemaArquivos = __importStar(require("fs"));
const caminho = __importStar(require("path"));
const tradutores_1 = require("@designliquido/delegua/tradutores");
const tradutor_assemblyscript_1 = require("@designliquido/delegua/tradutores/tradutor-assemblyscript");
const lexador_1 = require("@designliquido/delegua/lexador");
const avaliador_sintatico_1 = require("@designliquido/delegua/avaliador-sintatico");
const lexador_2 = require("@designliquido/visualg/lexador");
const avaliador_sintatico_2 = require("@designliquido/visualg/avaliador-sintatico");
const tradutores_2 = require("@designliquido/visualg/tradutores");
const avaliador_sintatico_javascript_1 = require("@designliquido/delegua/avaliador-sintatico/traducao/avaliador-sintatico-javascript");
const nucleo_comum_1 = require("./nucleo-comum");
const importador_1 = require("./importador");
const importador_javascript_1 = require("./importador/importador-javascript");
class NucleoTraducao extends nucleo_comum_1.NucleoComum {
    constructor(funcaoDeRetorno = null, funcaoDeRetornoMesmaLinha = null) {
        super();
        this.comandoTraducao = '';
        this.extensoes = {
            assemblyscript: '.as',
            delegua: '.delegua',
            javascript: '.js',
            js: '.js',
            alg: '.alg',
            visualg: '.alg',
            python: '.py',
            py: '.py',
        };
        this.arquivosAbertos = {};
        this.conteudoArquivosAbertos = {};
        this.funcaoDeRetorno = funcaoDeRetorno || console.log;
        // `process.stdout.write.bind(process.stdout)` é necessário por causa de 
        // https://stackoverflow.com/questions/28874665/node-js-cannot-read-property-defaultencoding-of-undefined
        this.funcaoDeRetornoMesmaLinha = funcaoDeRetornoMesmaLinha || process.stdout.write.bind(process.stdout);
    }
    iniciarTradutor(comandoTraducao) {
        switch (comandoTraducao) {
            case 'delegua-para-assemblyscript':
            case 'delegua-para-as':
                this.importador = new importador_1.Importador(new lexador_1.Lexador(false), this.arquivosAbertos, this.conteudoArquivosAbertos, false);
                this.avaliadorSintatico = new avaliador_sintatico_1.AvaliadorSintatico();
                this.tradutor = new tradutor_assemblyscript_1.TradutorAssemblyScript();
                break;
            case 'delegua-para-js':
            case 'delegua-para-javascript':
                this.importador = new importador_1.Importador(new lexador_1.Lexador(false), this.arquivosAbertos, this.conteudoArquivosAbertos, false);
                this.avaliadorSintatico = new avaliador_sintatico_1.AvaliadorSintatico();
                this.tradutor = new tradutores_1.TradutorJavaScript();
                break;
            case 'delegua-para-py':
            case 'delegua-para-python':
                this.importador = new importador_1.Importador(new lexador_1.Lexador(false), this.arquivosAbertos, this.conteudoArquivosAbertos, false);
                this.avaliadorSintatico = new avaliador_sintatico_1.AvaliadorSintatico();
                this.tradutor = new tradutores_1.TradutorPython();
                break;
            case 'js-para-delegua':
            case 'javascript-para-delegua':
                this.importador = new importador_javascript_1.ImportadorJavaScript();
                this.avaliadorSintatico = new avaliador_sintatico_javascript_1.AvaliadorSintaticoJavaScript();
                this.tradutor = new tradutores_1.TradutorReversoJavaScript();
                break;
            case 'alg-para-delegua':
            case 'visualg-para-delegua':
                this.importador = new importador_1.Importador(new lexador_2.LexadorVisuAlg(), this.arquivosAbertos, this.conteudoArquivosAbertos, false);
                this.avaliadorSintatico = new avaliador_sintatico_2.AvaliadorSintaticoVisuAlg();
                this.tradutor = new tradutores_2.TradutorReversoVisuAlg();
                break;
            default:
                throw new Error(`Tradutor '${comandoTraducao}' não implementado.`);
        }
    }
    /**
     * Realiza a tradução do arquivo passado como parâmetro no comando de execução.
     * @param caminhoRelativoArquivo O caminho do arquivo.
     * @param gerarArquivoSaida Se o resultado da tradução deve ser escrito em arquivo.
     *                          Se verdadeiro, os arquivos de saída são escritos no mesmo diretório
     *                          do arquivo passado no primeiro parâmetro.
     */
    traduzirArquivo(caminhoRelativoArquivo, gerarArquivoSaida) {
        var _a;
        const caminhoAbsolutoPrimeiroArquivo = caminho.resolve(caminhoRelativoArquivo);
        const novoDiretorioBase = caminho.dirname(caminhoAbsolutoPrimeiroArquivo);
        this.importador.diretorioBase = novoDiretorioBase;
        const retornoImportador = this.importador.importar(caminhoRelativoArquivo, true);
        let resultado = null;
        if (this.afericaoErrosLexador(retornoImportador.retornoLexador)) {
            process.exit(65); // Código para erro de avaliação antes da tradução
        }
        const retornoAvaliadorSintatico = this.avaliadorSintatico.analisar(retornoImportador.retornoLexador, retornoImportador.hashArquivo);
        resultado = this.tradutor.traduzir(retornoAvaliadorSintatico.declaracoes);
        if (gerarArquivoSaida) {
            const linguagem = ((_a = this.comandoTraducao) === null || _a === void 0 ? void 0 : _a.split('-')[2]) || '';
            const extensaoAlvo = this.extensoes[linguagem];
            if (extensaoAlvo) {
                ['.delegua', '.js', '.alg'].map((extensao) => {
                    if (caminhoAbsolutoPrimeiroArquivo.includes(extensao)) {
                        sistemaArquivos.writeFile(caminhoAbsolutoPrimeiroArquivo.replace(extensao, `${extensaoAlvo}`), resultado, (erro) => {
                            if (erro)
                                throw erro;
                        });
                        return;
                    }
                });
            }
        }
        this.funcaoDeRetorno(resultado);
    }
}
exports.NucleoTraducao = NucleoTraducao;
