// @designliquido/mapler/lexador/palavras-reservadas.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.palavrasReservadas = void 0;
const lexico_regular_1 = __importDefault(require("../tipos-de-simbolos/lexico-regular"));
exports.palavrasReservadas = {
    ate: lexico_regular_1.default.ATE,
    cadeia: lexico_regular_1.default.CADEIA,
    caractere: lexico_regular_1.default.CARACTERE,
    de: lexico_regular_1.default.DE,
    e: lexico_regular_1.default.E,
    enquanto: lexico_regular_1.default.ENQUANTO,
    entao: lexico_regular_1.default.ENTAO,
    escrever: lexico_regular_1.default.ESCREVER,
    faca: lexico_regular_1.default.FACA,
    falso: lexico_regular_1.default.FALSO,
    fim: lexico_regular_1.default.FIM,
    inicio: lexico_regular_1.default.INICIO,
    inteiro: lexico_regular_1.default.INTEIRO,
    ler: lexico_regular_1.default.LER,
    logico: lexico_regular_1.default.LOGICO,
    modulo: lexico_regular_1.default.MODULO,
    nao: lexico_regular_1.default.NEGACAO,
    ou: lexico_regular_1.default.OU,
    para: lexico_regular_1.default.PARA,
    passo: lexico_regular_1.default.PASSO,
    real: lexico_regular_1.default.REAL,
    repita: lexico_regular_1.default.REPITA,
    se: lexico_regular_1.default.SE,
    senao: lexico_regular_1.default.SENAO,
    variaveis: lexico_regular_1.default.VARIAVEIS,
    verdadeiro: lexico_regular_1.default.VERDADEIRO,
    vetor: lexico_regular_1.default.VETOR,
};
