// @designliquido/visualg/lexador/palavras-reservadas.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.palavrasReservadas = void 0;
const lexico_regular_1 = __importDefault(require("../tipos-de-simbolos/lexico-regular"));
exports.palavrasReservadas = {
    abs: lexico_regular_1.default.METODO_BIBLIOTECA_GLOBAL,
    algoritmo: lexico_regular_1.default.ALGORITMO,
    aleatorio: lexico_regular_1.default.ALEATORIO,
    arccos: lexico_regular_1.default.METODO_BIBLIOTECA_GLOBAL,
    arcsen: lexico_regular_1.default.METODO_BIBLIOTECA_GLOBAL,
    arctan: lexico_regular_1.default.METODO_BIBLIOTECA_GLOBAL,
    ate: lexico_regular_1.default.ATE,
    até: lexico_regular_1.default.ATÉ,
    caracter: lexico_regular_1.default.CARACTER,
    caractere: lexico_regular_1.default.CARACTERE,
    caso: lexico_regular_1.default.CASO,
    cos: lexico_regular_1.default.METODO_BIBLIOTECA_GLOBAL,
    cotan: lexico_regular_1.default.METODO_BIBLIOTECA_GLOBAL,
    de: lexico_regular_1.default.DE,
    e: lexico_regular_1.default.E,
    enquanto: lexico_regular_1.default.ENQUANTO,
    entao: lexico_regular_1.default.ENTAO,
    então: lexico_regular_1.default.ENTÃO,
    escolha: lexico_regular_1.default.ESCOLHA,
    escreva: lexico_regular_1.default.ESCREVA,
    escreval: lexico_regular_1.default.ESCREVA_LINHA,
    exp: lexico_regular_1.default.METODO_BIBLIOTECA_GLOBAL,
    faca: lexico_regular_1.default.FACA,
    faça: lexico_regular_1.default.FAÇA,
    falso: lexico_regular_1.default.FALSO,
    fimalgoritmo: lexico_regular_1.default.FIM_ALGORITMO,
    fimenquanto: lexico_regular_1.default.FIM_ENQUANTO,
    fimescolha: lexico_regular_1.default.FIM_ESCOLHA,
    fimfuncao: lexico_regular_1.default.FIM_FUNCAO,
    fimfunção: lexico_regular_1.default.FIM_FUNÇÃO,
    fimpara: lexico_regular_1.default.FIM_PARA,
    fimprocedimento: lexico_regular_1.default.FIM_PROCEDIMENTO,
    fimregistro: lexico_regular_1.default.FIM_REGISTRO,
    fimrepita: lexico_regular_1.default.FIM_REPITA,
    fimse: lexico_regular_1.default.FIM_SE,
    funcao: lexico_regular_1.default.FUNCAO,
    função: lexico_regular_1.default.FUNÇÃO,
    grauprad: lexico_regular_1.default.METODO_BIBLIOTECA_GLOBAL,
    inicio: lexico_regular_1.default.INICIO,
    int: lexico_regular_1.default.METODO_BIBLIOTECA_GLOBAL,
    inteiro: lexico_regular_1.default.INTEIRO,
    interrompa: lexico_regular_1.default.INTERROMPA,
    leia: lexico_regular_1.default.LEIA,
    limpatela: lexico_regular_1.default.LIMPA_TELA,
    log: lexico_regular_1.default.METODO_BIBLIOTECA_GLOBAL,
    logn: lexico_regular_1.default.METODO_BIBLIOTECA_GLOBAL,
    logico: lexico_regular_1.default.LOGICO,
    mod: lexico_regular_1.default.MODULO,
    nao: lexico_regular_1.default.NEGACAO, // Exceção
    on: lexico_regular_1.default.ON,
    off: lexico_regular_1.default.OFF,
    ou: lexico_regular_1.default.OU,
    outrocaso: lexico_regular_1.default.OUTRO_CASO,
    para: lexico_regular_1.default.PARA,
    passo: lexico_regular_1.default.PASSO,
    pi: lexico_regular_1.default.METODO_BIBLIOTECA_GLOBAL,
    procedimento: lexico_regular_1.default.PROCEDIMENTO,
    quad: lexico_regular_1.default.METODO_BIBLIOTECA_GLOBAL,
    radpgrau: lexico_regular_1.default.METODO_BIBLIOTECA_GLOBAL,
    raizq: lexico_regular_1.default.METODO_BIBLIOTECA_GLOBAL,
    rand: lexico_regular_1.default.METODO_BIBLIOTECA_GLOBAL,
    randi: lexico_regular_1.default.METODO_BIBLIOTECA_GLOBAL,
    real: lexico_regular_1.default.REAL,
    registro: lexico_regular_1.default.REGISTRO,
    repita: lexico_regular_1.default.REPITA,
    retorne: lexico_regular_1.default.RETORNE,
    se: lexico_regular_1.default.SE,
    sen: lexico_regular_1.default.METODO_BIBLIOTECA_GLOBAL,
    senao: lexico_regular_1.default.SENAO,
    senão: lexico_regular_1.default.SENÃO,
    tan: lexico_regular_1.default.METODO_BIBLIOTECA_GLOBAL,
    tipo: lexico_regular_1.default.TIPO,
    var: lexico_regular_1.default.VAR,
    verdadeiro: lexico_regular_1.default.VERDADEIRO,
    vetor: lexico_regular_1.default.VETOR,
    xou: lexico_regular_1.default.XOU,
};
