// @designliquido/visualg/bibliotecas/caracteres.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.asc = asc;
exports.carac = carac;
exports.caracpnum = caracpnum;
exports.compr = compr;
exports.copia = copia;
exports.maiusc = maiusc;
exports.minusc = minusc;
exports.numpcarac = numpcarac;
exports.pos = pos;
function asc(interpretador, valor) {
    const valorResolvido = valor.hasOwnProperty('valor') ? valor.valor : valor;
    return Promise.resolve(String(valorResolvido).charCodeAt(0));
}
function carac(interpretador, valor) {
    return Promise.resolve(String.fromCharCode(valor));
}
function caracpnum(interpretador, valor) {
    const valorResolvido = valor.hasOwnProperty('valor') ? valor.valor : valor;
    return Promise.resolve(Number(valorResolvido));
}
function compr(interpretador, valor) {
    const valorResolvido = valor.hasOwnProperty('valor') ? valor.valor : valor;
    return Promise.resolve(String(valorResolvido).length);
}
function copia(interpretador, valor, inicio, fim) {
    return Promise.resolve(valor.substring(inicio, inicio + fim));
}
function maiusc(interpretador, valor) {
    const valorResolvido = valor.hasOwnProperty('valor') ? valor.valor : valor;
    return Promise.resolve(String(valorResolvido).toUpperCase());
}
function minusc(interpretador, valor) {
    return Promise.resolve(valor.toLowerCase());
}
function numpcarac(interpretador, valor) {
    return Promise.resolve(String(valor));
}
function pos(interpretador, busca, valor) {
    const valorResolvido = valor.hasOwnProperty('valor') ? valor.valor : valor;
    return Promise.resolve(String(valorResolvido).indexOf(busca) + 1);
}
