// @designliquido/visualg/interpretador/estruturas/visualg-funcao.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VisuAlgFuncao = void 0;
const estruturas_1 = require("@designliquido/delegua/estruturas");
const quebras_1 = require("@designliquido/delegua/quebras");
/**
 * Diferentemente de `DeleguaFuncao`, a forma de VisuAlg de trabalhar com referências usa
 * como base o nome do parâmetro, e não o nome do argumento, como é em Delégua.
 */
class VisuAlgFuncao extends estruturas_1.DeleguaFuncao {
    async chamar(visitante, argumentos) {
        const ambiente = this.resolverAmbiente(argumentos);
        if (this.instancia !== undefined) {
            ambiente.valores['isto'] = {
                valor: this.instancia,
                tipo: 'objeto',
                imutavel: false,
            };
        }
        visitante.proximoEscopo = 'funcao';
        const retornoBloco = await visitante.executarBloco(this.declaracao.corpo, ambiente);
        const referencias = this.declaracao.parametros
            .map((p, indice) => {
            if (p.referencia) {
                return {
                    indice: indice,
                    parametro: p,
                };
            }
        })
            .filter((r) => r);
        const pilha = visitante.pilhaEscoposExecucao;
        for (let referencia of referencias) {
            let argumentoReferencia = ambiente.valores[referencia.parametro.nome.lexema];
            // TODO: Lógica implementada para o VisuAlg. 
            pilha.atribuirVariavel({
                lexema: argumentos[referencia.indice].nome
            }, argumentoReferencia.valor);
        }
        if (retornoBloco instanceof quebras_1.RetornoQuebra) {
            return retornoBloco.valor;
        }
        if (this.eInicializador) {
            return this.instancia;
        }
        return retornoBloco;
    }
    funcaoPorMetodoDeClasse(instancia) {
        return new VisuAlgFuncao(this.nome, this.declaracao, instancia, this.eInicializador);
    }
}
exports.VisuAlgFuncao = VisuAlgFuncao;
