// @designliquido/portugol-studio/interpretador/interpretador-portugol-studio-com-depuracao.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.InterpretadorPortugolStudioComDepuracao = void 0;
const interpretador_com_depuracao_1 = require("@designliquido/delegua/interpretador/interpretador-com-depuracao");
const pilha_escopos_execucao_portugol_studio_1 = require("./pilha-escopos-execucao-portugol-studio");
const comum = __importStar(require("./comum"));
class InterpretadorPortugolStudioComDepuracao extends interpretador_com_depuracao_1.InterpretadorComDepuracao {
    constructor(diretorioBase, funcaoDeRetorno = null, funcaoDeRetornoMesmaLinha = null, funcaoLimpa = null) {
        super(diretorioBase, funcaoDeRetorno, funcaoDeRetornoMesmaLinha);
        this.funcaoLimpa = () => {
            console.log('Função "limpa()" não está ligada a uma interface de entrada e saída.');
        };
        if (funcaoLimpa !== null) {
            this.funcaoLimpa = funcaoLimpa;
        }
        this.mensagemPrompt = '> ';
        this.pilhaEscoposExecucao = new pilha_escopos_execucao_portugol_studio_1.PilhaEscoposExecucaoPortugolStudio();
    }
    async visitarExpressaoLimpa(expressao) {
        this.funcaoLimpa();
        return Promise.resolve();
    }
    /**
     * Execução de uma escrita na saída padrão, sem quebras de linha, e sem remoção de espaços ao final.
     * @param declaracao A declaração.
     * @returns Sempre nulo, por convenção de visita.
     */
    async visitarDeclaracaoEscrevaMesmaLinha(declaracao) {
        try {
            const formatoTexto = await comum.avaliarArgumentosEscreva(this, declaracao.argumentos);
            this.funcaoDeRetornoMesmaLinha(formatoTexto);
            return null;
        }
        catch (erro) {
            this.erros.push({
                erroInterno: erro,
                linha: declaracao.linha,
                hashArquivo: declaracao.hashArquivo,
            });
        }
    }
    async visitarDeclaracaoImportar(declaracao) {
        return comum.visitarExpressaoImportarComum(declaracao);
    }
    /**
     * Execução da leitura de valores da entrada configurada no
     * início da aplicação.
     * @param expressao Expressão do tipo Leia
     * @returns Promise com o resultado da leitura.
     */
    async visitarExpressaoLeia(expressao) {
        return comum.visitarExpressaoLeiaComum(this, this.interfaceEntradaSaida, expressao);
    }
    async visitarExpressaoMatriz(expressao) {
        return comum.visitarExpressaoMatrizComum(this, expressao);
    }
    /**
     * No Portugol Studio, como o bloco de execução da função `inicio` é criado
     * pelo avaliador sintático, precisamos ter uma forma aqui de avançar o
     * primeiro bloco pós execução de comando, seja ele qual for.
     */
    avancarPrimeiroEscopoAposInstrucao() {
        const escopoUm = this.pilhaEscoposExecucao.naPosicao(1);
        if (!escopoUm)
            return;
        escopoUm.declaracaoAtual = escopoUm.declaracoes.length;
    }
    async instrucaoContinuarInterpretacao(escopo) {
        const retornoExecucao = await super.instrucaoContinuarInterpretacao(escopo);
        this.avancarPrimeiroEscopoAposInstrucao();
        return retornoExecucao;
    }
    async instrucaoPasso(escopo) {
        const retornoExecucaoPasso = await super.instrucaoPasso(escopo);
        this.avancarPrimeiroEscopoAposInstrucao();
        return retornoExecucaoPasso;
    }
}
exports.InterpretadorPortugolStudioComDepuracao = InterpretadorPortugolStudioComDepuracao;
