// @designliquido/potigol/interpretador/interpretador-potigol-com-depuracao.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.InterpretadorPotigolComDepuracao = void 0;
const interpretador_com_depuracao_1 = require("@designliquido/delegua/interpretador/interpretador-com-depuracao");
const comum = __importStar(require("./comum"));
class InterpretadorPotigolComDepuracao extends interpretador_com_depuracao_1.InterpretadorComDepuracao {
    constructor(diretorioBase, funcaoDeRetorno = null, funcaoDeRetornoMesmaLinha = null) {
        super(diretorioBase, funcaoDeRetorno, funcaoDeRetornoMesmaLinha);
        this.expandirPropriedadesDeObjetosEmEspacoVariaveis = true;
        this.regexInterpolacao = /{(.*?)}/g;
        comum.carregarBibliotecaGlobal(this.pilhaEscoposExecucao);
    }
    visitarDeclaracaoLeiaInteiros(declaracao) {
        return comum.visitarExpressaoLeiaMultiplo(this, declaracao);
    }
    visitarDeclaracaoLeiaReais(declaracao) {
        return comum.visitarExpressaoLeiaMultiplo(this, declaracao);
    }
    visitarDeclaracaoLeiaTextos(declaracao) {
        return comum.visitarExpressaoLeiaMultiplo(this, declaracao);
    }
    visitarDeclaracaoLeiaInteiro(declaracao) {
        return comum.visitarExpressaoLeia(this, declaracao);
    }
    visitarDeclaracaoLeiaReal(declaracao) {
        return comum.visitarExpressaoLeia(this, declaracao);
    }
    visitarDeclaracaoLeiaTexto(declaracao) {
        return comum.visitarExpressaoLeia(this, declaracao);
    }
    async resolverInterpolacoes(textoOriginal, linha) {
        return comum.resolverInterpolacoes(this, textoOriginal, linha);
    }
    retirarInterpolacao(texto, variaveis) {
        return comum.retirarInterpolacao(texto, variaveis);
    }
    async visitarDeclaracaoReatribuicaoVariavel(expressao) {
        return comum.visitarDeclaracaoReatribuicaoVariavel(this, expressao);
    }
    async visitarExpressaoAcessoMetodoOuPropriedade(expressao) {
        return comum.visitarExpressaoAcessoMetodoOuPropriedade(this, expressao);
    }
    async visitarExpressaoBinaria(expressao) {
        return comum.visitarExpressaoBinaria(this, expressao);
    }
    async visitarExpressaoQualTipo(expressao) {
        return comum.visitarExpressaoQualTipo(this, expressao);
    }
    async avaliarArgumentosEscreva(argumentos) {
        return comum.avaliarArgumentosEscreva(this, argumentos.length > 0 ? argumentos[0] : undefined);
    }
}
exports.InterpretadorPotigolComDepuracao = InterpretadorPotigolComDepuracao;
