// @designliquido/portugol-studio/tipos-de-simbolos/lexico-regular.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = {
    ADICAO: 'ADICAO',
    BIBLIOTECA: 'BIBLIOTECA',
    CADEIA: 'CADEIA',
    CARACTER: 'CARACTER',
    CASO: 'CASO',
    CHAVE_ESQUERDA: 'CHAVE_ESQUERDA',
    CHAVE_DIREITA: 'CHAVE_DIREITA',
    COLCHETE_ESQUERDO: 'COLCHETE_ESQUERDO',
    COLCHETE_DIREITO: 'COLCHETE_DIREITO',
    COMENTARIO: 'COMENTARIO',
    CONSTANTE: 'CONSTANTE',
    CONTRARIO: 'CONTRARIO',
    DECREMENTAR: 'DECREMENTAR',
    DIFERENTE: 'DIFERENTE',
    DIVISAO: 'DIVISAO',
    DIVISAO_IGUAL: 'DIVISAO_IGUAL',
    DIVISAO_INTEIRA: 'DIVISAO_INTEIRA',
    DOIS_PONTOS: 'DOIS_PONTOS',
    E: 'E',
    ENQUANTO: 'ENQUANTO',
    ESCOLHA: 'ESCOLHA',
    ESCREVA: 'ESCREVA',
    FACA: 'FACA',
    FALSO: 'FALSO',
    FUNCAO: 'FUNCAO',
    IDENTIFICADOR: 'IDENTIFICADOR',
    IGUAL: 'IGUAL',
    IGUAL_IGUAL: 'IGUAL_IGUAL',
    INCREMENTAR: 'INCREMENTAR',
    INCLUA: 'INCLUA',
    INTEIRO: 'INTEIRO',
    LEIA: 'LEIA',
    LIMPA: 'LIMPA',
    LINHA_COMENTARIO: 'LINHA_COMENTARIO',
    LOGICO: 'LOGICO',
    MAIOR: 'MAIOR',
    MAIOR_IGUAL: 'MAIOR_IGUAL',
    MAIS_IGUAL: 'MAIS_IGUAL',
    MENOR: 'MENOR',
    MENOR_IGUAL: 'MENOR_IGUAL',
    MENOS_IGUAL: 'MENOS_IGUAL',
    MODULO: 'MODULO',
    MULTIPLICACAO: 'MULTIPLICACAO',
    MULTIPLICACAO_IGUAL: 'MULTIPLICACAO_IGUAL',
    NEGACAO: 'NEGACAO',
    OU: 'OU',
    PARA: 'PARA',
    PARE: 'PARE',
    PARENTESE_ESQUERDO: 'PARENTESE_ESQUERDO',
    PARENTESE_DIREITO: 'PARENTESE_DIREITO',
    PONTO: 'PONTO',
    PONTO_E_VIRGULA: 'PONTO_E_VIRGULA',
    PROGRAMA: 'PROGRAMA',
    RETORNE: 'RETORNE',
    REAL: 'REAL',
    SETA: 'SETA',
    SUBTRACAO: 'SUBTRACAO',
    VIRGULA: 'VIRGULA',
    SE: 'SE',
    SENAO: 'SENAO',
    VAZIO: 'VAZIO',
    VERDADEIRO: 'VERDADEIRO',
};
